prov/efa/docs/efa_rdm_protocol_v4.md

```diff
@@ -0,0 +1,1318 @@
+   extra/feature request status. Therefore, it can only use the baseline features to
+   communicate with a peer, which means it will send REQ packets (section 3.1) to the peer
+   to initialize a communication.
+2. Upon receiving the 1st REQ packet from a peer, an endpoint will send back a handshake
+   packet, which contains the endpoint's capability information.
+3. Upon receiving the handshake packet, an endpoint will know the peer's extra feature/request
+   status.
+
+Regarding extra feature, if the peer support the extra feature the endpoint want to use,
+the endpoint can start using the extra feature. Otherwise, one of the following should happen:
+
+- a. the communication continues without using the extra feature/request, though
+   the performance may be sub-optimal. For example, if the peer does not support
+   the extra feature "RDMA read based data transfer", the endpoint can choose to
+   use baseline features to carry on the communication, though the performance will
+   be sub-optimal (section 4.1).
+
+- b. the requester of the communication aborts the communication and return an error
+   to the application. For example, if application requires delivery complete, but
+   the peer does not support it (this can happen when endpoint is using libfabric 1.12,
+   but the peer is using libfabric 1.10), the requester need to return an error to the
+   application (section 4.2)
+
+Regarding extra request, if an endpoint can support an extra request the peer has, it
+should comply the request. Otherwise, it can ignore the request. Peer should be do
+one of the following:
+
+- a. carry on the communication without using the extra request. (see section 4.4
+     for example)
+
+- b. abort the communication and return an error to application. (see section 4.3
+     for example)
+
+For example, if sender is using libfabric 1.10, and receiver is using 1.14. Receiver
+will have the extra request "connid header", but sender does not support it. In this
+case, sender should not do anything special, it is receiver's responsibility to react
+accordingly. (section 4.4)
+
+This concludes the workflow of the handshake sub-protocol.
+
+The binary format of a HANDSHAKE packet is listed in table 2.2.
+
+Table: 2.2 binary format of the HANDSHAKE packet
+
+| Name      | Length (bytes) | type | C language type |
+|-|-|-|-|
+| `type`    | 1 | integer | `uint8_t`  |
+| `version` | 1 | integer | `uint8_t`  |
+| `flags`   | 2 | integer | `uint16_t` |
+| `nex_p3`  | 4 | integer | `uint32_t` |
+| `exinfo`  | `8 * (nex_p3 - 3)` | integer array | `uint64_t[]` |
+
+The first 4 bytes (3 fields: `type`, `version`, `flags`) is the EFA RDM base header (section 1.3).
+
+Immediately after the base header, there are 2 fields `nex_p3` and `exinfo`. 
+
+The field `exinfo` is an array of 8 byte integers, which stores the capability of an endpoint.
+
+As mentioned before, each extra feature/request was assigned an ID when it was introduced to protocol v4.
```

Comment:
This is just a bitmap, right? I think I'd call it that in the type field above and simplify this paragraph.